### PR TITLE
Added Multi-Tenant Capabilities for Different eCHIS Instances. [uses X-OpenHIM-ClientID in headers]

### PR DESCRIPTION
**Pull request by karimkkanji**
**Description**
Link to Github Ticket: https://github.com/moh-kenya/echis-2.0-integration/issues/121

**What does this PR do?:**
This Pull request introduces the use of X-OpenHIM-ClientID headers coming from OpenHIM as here: 

image

This header is passed when a client is setup and the same client is used in echis as seen here:

image

**How was the solution implemented?**
1. Adding an .env file that contains all credentials for the various instances.
2. Created a function in the mediator to handle the request header and sets the instance in a global variable.
3. Changed CHT object to a function to allow it to pick values dynamically from env.
4. Route to the right eCHIS instance based on the picked values from env.

**How can the changes be tested?**
The changes can be deployed to staging to test. By default echis has been handled. This can be setup locally for both echis and mediator. 

Ensure the env file is updated accordingly.

### CODE REVIEW

File: src/utils/globalInstance.js
Comment (freddieptf) [inline src/utils/globalInstance.js:1]: I don't think global state is a good idea here. This can be done via a request middleware

File: env.template
Comment (freddieptf) [inline env.template:1]: No need to include all this fields here as they are environment dependent

File: src/controllers/aggregate.js [hunk 1]
```diff
@@ -33,7 +33,14 @@ const getMoh515Data = async (_, response) => {
       throw error;
     }
     const result = results.rows;
-    response.send(result);
+    response.setHeader("Content-Type", "application/json");
```
Comment (freddieptf) [inline src/controllers/aggregate.js:36]: Is this change relevant to the issue?

File: src/controllers/client.js [hunk 2]
```diff
@@ -194,20 +194,22 @@ axiosInstance.interceptors.response.use(
   }
 );
 
-const echisAxiosInstance = axios.create({
-  baseURL: CHT.url,
+const echisAxiosInstance =()=>{
```
Comment (freddieptf) [inline src/controllers/client.js:197]: we can use the default axios instance with different request config i.e auth, base url etc. There are utility functions you can use for this in https://github.com/moh-kenya/echis-2.0-integration/blob/main/src/utils/echis.js

**Review by freddieptf — CHANGES_REQUESTED** [review]

**Comment by freddieptf** [comment]
Almost there but lets rebase this first @karimkkanji, there are some changes on the main branch that you can benefit from while implementing this

File: config.js [hunk 3]
```diff
@@ -35,10 +37,12 @@ const FHIR = {
   url: process.env.FHIR_URL,
 };
 
-const CHT = {
-  url: process.env.CHT_URL,
-  username: process.env.CHT_USERNAME,
-  password: process.env.CHT_PASSWORD,
+const CHT =()=>{
```
Comment (freddieptf) [inline config.js:40]: lets pass the instance as a param to this function

**Comment by karimkkanji** [comment]
> Almost there but lets rebase this first @karimkkanji, there are some changes on the main branch that you can benefit from while implementing this

Rebased. Working on the other reviews

File: env.template
Comment (karimkkanji) [reply env.template:1]: Please Elaborate. The mediator will contain all credentials for all instances. This template is a starting point or what will eventually look like in the env file.

File: src/controllers/aggregate.js (hunk 1 above)
Comment (karimkkanji) [reply src/controllers/aggregate.js:36]: This was pulled in from PR 125

File: env.template
Comment (freddieptf) [reply env.template:1]: It's a template, it should just have a small example of how to define the variables. Adding all these items here implies this template is up to date with the state of all deployments in an environment, which it will most likely not always be. We don't want devs blindly copy pasting this in the future and then wondering why something isn't working when they either need to update or add something to this

File: src/utils/globalInstance.js
Comment (freddieptf) [reply src/utils/globalInstance.js:1]: @karimkkanji did you work on this?

**Comment by freddieptf** [comment]
> The rebase has made the reviews outdated.

@karimkkanji yes but the comments are still relevant

File: src/utils/globalInstance.js
Comment (karimkkanji) [reply src/utils/globalInstance.js:1]: Updated this to use setInstance Middleware

File: config.js (hunk 3 above)
Comment (karimkkanji) [reply config.js]: Now passing Instance as a param.

File: env.template
Comment (karimkkanji) [reply env.template:1]: Updated the template with instructions on how to add each instance

File: src/middlewares/setInstance.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const { logger } = require("../utils/logger");
+
+module.exports = function (instanceValue) {
+    return setInstance=(req, res, next) => {
+        if (req.headers['x-openhim-clientid']) {
+          const clientId = req.headers['x-openhim-clientid'];
+          if (clientId === 'echis') {
+            res.locals.instanceValue = "STAGING";
```
Comment (freddieptf) [inline src/middlewares/setInstance.js:8]: Let's not define defaults here, the middleware should work with only what is passed in the request headers and fail if the header does not exist

File: src/middlewares/setInstance.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+          } else {
+            res.locals.instanceValue = clientId;
+          }
+        } else {
+          res.locals.instanceValue = "STAGING";
+        }
+        logger.information(`Setting instance value to: ${res.locals.instanceValue}`);
```
Comment (freddieptf) [inline src/middlewares/setInstance.js:15]: Let's try and only log informative errors for the sake of our signal to noise ratio, it's very hard to follow logs as is

File: src/routes/referral.js [hunk 6]
```diff
@@ -22,10 +21,9 @@ router.post("/community", async function (req, res) {
 });
 
 router.post("/facility", async function (req, res) {
-  setGlobalInstance(req);
-  const { status, serviceRequestId } = await createFacilityReferral(req.body);
+  const { status, serviceRequestId } = await createFacilityReferral(req.body,res);
   res.setHeader("Content-Type", "application/json");
-  res.status(status).send(
+  res.status(status || 500).send(
```
Comment (freddieptf) [inline src/routes/referral.js:26]: The function should return the status code instead of relying on a default

File: src/controllers/referral.js [hunk 7]
```diff
@@ -143,19 +146,20 @@ const createCommunityReferral = async (serviceRequest) => {
...
     const axiosInstance = axios.create({
-      baseURL: CHT().url,
+      baseURL: CHT(instanceValue).url,
       headers: {
         "Content-Type": "application/json",
       },
       auth: {
-        username: CHT().username,
-        password: CHT().password,
+        username: CHT(instanceValue).username,
```
Comment (freddieptf) [inline src/controllers/referral.js:161]: Let's reuse the result of calling the CHT function here

File: src/controllers/referral.js [hunk 8]
```diff
@@ -20,10 +20,11 @@ const {
   COMPLETED_SUCCESSFULLY,
 } = messages;
 
-const getSubjectUpi = async (echisClientId) => {
+const getSubjectUpi = async (echisClientId,instance) => {
```
Comment (freddieptf) [inline src/controllers/referral.js:23]: We can rearrange the order of params here to match other functions that expect and instance object

File: config.js [hunk 9]
```diff
@@ -37,11 +37,11 @@ const FHIR = {
   url: process.env.FHIR_URL,
 };
 
-const CHT =()=>{
+const CHT =(instance)=>{
```
Comment (freddieptf) [inline config.js:40]: This function can be renamed to something more appropriate

**Review by freddieptf — CHANGES_REQUESTED** [review]
Good Progress @karimkkanji! I've left some more comments

File: src/controllers/client.js [hunk 10]
```diff
@@ -68,21 +68,23 @@ const createClientDetailsMismatchReport = (conf, echisClientId, fields) => {
...
-    res.status(code).send({
+    res.status(code || 500).send({
       data: {
         echis_doc: echisClientId,
         upi: clientNumber,
       }
     })
   };
+  const instance = res.locals.instanceValue;
+  const instanceObject = { instance: CHT(instance).url, user: CHT(instance).username, password: CHT(instance).password };
```
[the remark this answers is not in the document]
Comment (karimkkanji) [reply src/controllers/client.js:79]: True.

File: src/routes/referral.js (hunk 6 above)
Comment (karimkkanji) [reply src/routes/referral.js:26]: There is an issue with the passed statuses being undefined.

File: src/controllers/client.js [hunk 11]
```diff
@@ -68,21 +68,23 @@ const createClientDetailsMismatchReport = (conf, echisClientId, fields) => {
 
 async function assignEchisClientUPI(req, res) {
   const success = (code, echisClientId, clientNumber) => {
-    res.status(code).send({
+    res.status(code || 500).send({
```
[the remark this answers is not in the document]
Comment (freddieptf) [reply src/controllers/client.js:71]: Then we should be fixing that, what path is calling the success function without a code?

File: src/controllers/referral.js (hunk 8 above)
Comment (karimkkanji) [reply src/controllers/referral.js:23]: Should we have it come first?

File: src/controllers/client.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (karimkkanji) [reply src/controllers/client.js]: This is what is causing the 500 error from the server:

https://github.com/moh-kenya/echis-2.0-integration/blob/4a9c3adace6a140d1d6cb1e3d1a00509e3ccd158/src/controllers/referral.js#L141

Here is a hit using Postman:
image

File: src/controllers/referral.js (hunk 8 above)
Comment (karimkkanji) [reply src/controllers/referral.js]: Updated the functions.

File: src/middlewares/setInstance.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+const { logger } = require("../utils/logger");
+
+module.exports = function () {
+    return setInstance=(req, res, next) => {
+        if (req.headers['x-openhim-clientid']) {
+          const clientId = req.headers['x-openhim-clientid'];
+            res.locals.instanceValue = clientId;
+        } else {
+          res.status(500).send({ error: "Required Headers were not passed"})
```
Comment (freddieptf) [inline src/middlewares/setInstance.js:9]: We should be returning a [400](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400)

**Review by freddieptf — CHANGES_REQUESTED** [review]
Almost there @karimkkanji, Lets now test how this works with some of our existing testing instances

File: src/middlewares/setInstance.js (hunk 12 above)
Comment (karimkkanji) [reply src/middlewares/setInstance.js:9]: Updating.

File: src/utils/client.js [hunk 13]
```diff
@@ -164,7 +164,7 @@ function getMismatchedClientFields(echisClientDoc, crClientDoc) {
   matcherFields.forEach(key => {
     const actual = getPropByString(crPayload, key);
     const expected = getPropByString(crClientDoc, key);
-    if (actual.toUpperCase() !== expected.toUpperCase()) {
+    if (actual?.toUpperCase() !== expected?.toUpperCase()) {
```
Comment (freddieptf) [inline src/utils/client.js:167]: This just hides the problem. Can you confirm what field was missing? Is it related to https://github.com/moh-kenya/echis-2.0-integration/issues/126?

File: src/controllers/client.js [hunk 14]
```diff
@@ -117,7 +117,7 @@ async function assignEchisClientUPI(req, res) {
     success(201, res.locals.echisClient._id, clientNumber)
   } catch (err) {
     const errStruct = { instance: getCHTValuesFromEnv(instance).url, user: res.locals.echisClient._id, error: err.message };
-    const errString = `could not complete creating client ${JSON.stringify(errStruct)})`
+    const errString = `Could not complete creating client. Possible Missing Fields. ${JSON.stringify(errStruct)})`
```
Comment (freddieptf) [inline src/controllers/client.js:120]: `Possible Missing Fields` isn't the only reason the request could fail. The error `400` returned by the client registry API should serve as a hint for the reason. Does the API return the missing fields if there are some? If so then maybe using the error from CR would be more helpful

Comment (karimkkanji) [reply src/controllers/client.js:120]: The issue is some of the functions do not have a catch statement. Such as the createCRClient Function. https://github.com/moh-kenya/echis-2.0-integration/blob/512756642206300ad616c088808822361c5de3c2/src/controllers/client.js#L45

```
const createCRClient = (conf, echisClient) => {
  const payload = generateClientRegistryPayload(echisClient);
  return axiosInstance.post(`partners/registry`, payload, {
    headers: {
      "Content-Type": "application/json",
    },
  }).then((resp) => {
    return updateDoc(conf, echisClient._id, { upi: resp.data.clientNumber })
      .then(() => Promise.resolve(resp.data.clientNumber));
  }).catch((err)=>{
      logger.error(err.response.data.errors);
  })
}
```

Possiby as above and handling appropriately

Comment (freddieptf) [reply src/controllers/client.js:120]: > The issue is some of the functions do not have a catch statement.

That's by design, the caller of the function should catch the error. If the CR errors are in `err.response.data.errors`, can't you just use that where the function is called?

File: src/utils/client.js (hunk 13 above)
Comment (freddieptf) [reply src/utils/client.js:167]: I get that but in the error you shared it was failing because `expected` was undefined, correct? `expected` is a field from the CR client doc so that means `getPropByString` is returning undefined.

File: src/controllers/client.js (hunk 14 above)
Comment (karimkkanji) [reply src/controllers/client.js]: Reverted the changes.

File: src/utils/client.js (hunk 13 above)
Comment (karimkkanji) [reply src/utils/client.js]: Reverted the changes.

Comment (karimkkanji) [reply src/utils/client.js]: The mediator has updated the missing fields on eCHIS as shown below and works in the expected behavior. The problem is the service dies and returns a 502 Bad Gateway. 

image

Comment (freddieptf) [reply src/utils/client.js]: > The problem is the service dies and returns a 502 Bad Gateway.

This is what we're trying to resolve, it will crash again if we don't know what caused the initial error. Can you link me one example transaction i can use to replicate the failure you were getting in the demo?

File: src/controllers/client.js [hunk 15]
```diff
@@ -20,39 +20,47 @@ const tryClientRegistryReauth = async (error) => {
...
-  return axiosInstance.post(`partners/registry`, payload, {
-    headers: {
-      "Content-Type": "application/json",
-    },
-  }).then((resp) => {
-    return updateDoc(conf, echisClient._id, { upi: resp.data.clientNumber })
-      .then(() => Promise.resolve(resp.data.clientNumber));
-  });
-}
+  return axiosInstance
```
Comment (freddieptf) [inline src/controllers/client.js:52]: Lots of reformatting going on here, doesn't it look a little less readable now?

Comment (karimkkanji) [reply src/controllers/client.js:52]: Prettier doing its thing :(

File: src/controllers/client.js [hunk 16]
```diff
@@ -72,56 +80,89 @@ async function assignEchisClientUPI(req, res) {
...
+    return; // nothing more to do here
   }
   // echis client did not exist in the registry, lets create them
   try {
-    const clientNumber = await createCRClient({ instance: CHT.url, user: CHT.username, password: CHT.password }, res.locals.echisClient);
-    success(201, res.locals.echisClient._id, clientNumber)
+    const clientNumber = await createCRClient(
+      instanceObject,
+      res.locals.echisClient
+    ).catch((error) => {
```
Comment (freddieptf) [inline src/controllers/client.js:146]: do not catch the error here, we have a try catch for that. Build the error message in the catch block

File: src/controllers/client.js (hunk 15 above)
Comment (karimkkanji) [reply src/controllers/client.js:52]: Prettier doing its thing :(

File: src/controllers/client.js [hunk 17]
```diff
@@ -72,56 +80,89 @@ async function assignEchisClientUPI(req, res) {
...
+      res.status(400).send({ error: error.response.data.errors });
```
[the remark this answers is not in the document]
Comment (karimkkanji) [reply src/controllers/client.js]: @freddieptf Thank you for this. This was the root cause. Seeing all requests being returned as completed.

File: src/utils/client.js (hunk 13 above)
Comment (karimkkanji) [reply src/utils/client.js]: Resolved it under this review:

https://github.com/moh-kenya/echis-2.0-integration/pull/125/#discussion_r1356916315
